From fdcd3f666190de7bc685ae4975210c1247849259 Mon Sep 17 00:00:00 2001
From: Jeff Mahoney <jeffm@suse.com>
Date: Tue, 12 Jan 2016 20:45:21 -0500
Subject: [PATCH 07/24] py-symbol: use Py_RETURN_NONE instead of open coding it

---
 gdb/python/py-symbol.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/gdb/python/py-symbol.c b/gdb/python/py-symbol.c
index d683505c8e5..39ba7d190a4 100644
--- a/gdb/python/py-symbol.c
+++ b/gdb/python/py-symbol.c
@@ -74,8 +74,7 @@ sympy_get_type (PyObject *self, void *closure)
 
   if (SYMBOL_TYPE (symbol) == NULL)
     {
-      Py_INCREF (Py_None);
-      return Py_None;
+      Py_RETURN_NONE;
     }
 
   return type_to_type_object (SYMBOL_TYPE (symbol));
@@ -466,8 +465,7 @@ gdbpy_lookup_global_symbol (PyObject *self, PyObject *args, PyObject *kw)
     }
   else
     {
-      sym_obj = Py_None;
-      Py_INCREF (Py_None);
+      Py_RETURN_NONE;
     }
 
   return sym_obj;
-- 
2.26.2

