From a56b41ce7dd5e8fd249d206ba09e6e505a5ecdf9 Mon Sep 17 00:00:00 2001
From: Jeff Mahoney <jeffm@suse.com>
Date: Sat, 13 Apr 2019 23:05:29 -0400
Subject: [PATCH 22/24] terrible symfile hack

Acked-by: Jeff Mahoney <jeffm@suse.com>
---
 gdb/symfile.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gdb/symfile.c b/gdb/symfile.c
index 7d8d84abf80..dfc40744cd4 100644
--- a/gdb/symfile.c
+++ b/gdb/symfile.c
@@ -660,7 +660,7 @@ default_symfile_offsets (struct objfile *objfile,
 	if (bfd_section_vma (cur_sec) != 0)
 	  break;
 
-      if (cur_sec == NULL)
+      if (1 || cur_sec == NULL)
 	{
 	  CORE_ADDR *offsets = objfile->section_offsets->offsets;
 
-- 
2.26.2

